spacy/tokens/doc.pyx

```diff
@@ -1462,6 +1462,146 @@ cdef class Doc:
+            token_annotations[HEAD]["values"] = [
+                head - i if head is not None else 0 for i, head in enumerate(token_annotations[HEAD]["values"])
+            ]
+        if token_annotations[DEP]["values"] is not None:
+            deps = []
+            for dep in token_annotations[DEP]["values"]:
+                if dep is None:
+                    raise ValueError(Errors)
+                deps.append(dep)
+            token_annotations[DEP]["values"] = deps if len(deps) else None
+
+        if token_annotations[DEP]["values"] and not token_annotations[HEAD]["values"]:
+            token_annotations[HEAD]["values"] = [0] * len(token_annotations[DEP]["values"])
+        if token_annotations[HEAD]["values"] and not token_annotations[HEAD]["values"]:
+            raise ValueError(Errors.E1017)
+        if token_annotations[POS]["values"] is not None:
+            for pp in set(token_annotations[POS]["values"]):
+                if pp not in parts_of_speech.IDS:
+                    raise ValueError(Errors.E1021.format(pp=pp))
+
+        headings = []
+        annotations = [token_annotations[key]["values"] for key in annot_types]
+        possible_headings = [POS, HEAD, DEP, LEMMA, TAG, MORPH]
+        for a, annot in enumerate(annotations):
+            if annot is not None:
+                if len(annot) != len(words):
+                    raise ValueError(Errors.E189)
+                headings.append(possible_headings[a])
+
+        # If there are any other annotations, set them.
+        if headings:
+            attrs = self.to_array(headings)
+            j = 0
+
+            for annot in annotations:
+                if annot:
+                    if annot is not token_annotations[HEAD]["values"]:
+                        for i in range(len(annot)):
+                            if annot[i] is not None:
+                                self.vocab.strings.add(annot[i])
+
+                    if annot is token_annotations[HEAD]["values"]:
+                        for i in range(len(words)):
+                            if attrs.ndim == 1:
+                                attrs[i] = annot[i]
+                            else:
+                                attrs[i, j] = annot[i]
+                    elif annot is token_annotations[MORPH]["values"]:
+                        for i in range(len(words)):
+                            morph_key = self.vocab.morphology.add(token_annotations[MORPH]["values"][i])
+                            if attrs.ndim == 1:
+                                attrs[i] = morph_key
+                            else:
+                                attrs[i, j] = morph_key
+                    else:
+                        for i in range(len(words)):
+                            if attrs.ndim == 1:
+                                attrs[i] = self.vocab.strings[annot[i]]
+                            else:
+                                attrs[i, j] = self.vocab.strings[annot[i]]
```

Comment:
You do need `self.vocab.strings.add(annot[i])` here instead. It's confusing, but `self.vocab.strings[x]` only gets the hash, it doesn't add it to the string store.

Reply:
So 
```python
if attrs.ndim == 1:
    attrs[i] = self.vocab.strings[annot[i]]
else:
    attrs[i, j] = self.vocab.strings[annot[i]]
self.vocab.strings.add(annot[i])
```
despite `attrs` being passed to `.from_array()` immediately after?

Reply:
It can look more like `morph_key` above? The `attrs` are passed as `uint64` and if the strings weren't already added to the string store there's no way to recover them.

Reply:
(And it's not your fault, but this use of `attrs` as a variable name here is extremely confusing.)

Reply:
Handled the vocabulary addition like `morph_key` now:
```python
vocab_key = self.vocab.strings.add(annot[i])
if array.ndim == 1:
    array[i] = vocab_key
else:
    array[i, j] = vocab_key
```
Is that what you meant?